platform-sdk/docs/proposals/metric-labels/README.md

```diff
@@ -0,0 +1,341 @@
+
+interface Metrics {
+    
+    // old methods:
+    // Object getValue(@NonNull String category, @NonNull String name);
+    // void remove(@NonNull String category, @NonNull String name);
+    
+    @Nullable
+    Object getValue(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
+    
+    void remove(@NonNull String category, @NonNull String name, @NonNull Set<Label> labels);
+}
+```
+
+Since the old methods are only used in tests or demos today we can do that change easily.
+
+> [!NOTE]  
+Since most of the filter methods that the `Metrics` interface defines today are not used we will not introduce any additional filter methods in this proposal.
+If we see that additional methods are needed we can add them later.
+
+We will have labels that are global for a `Metrics` instance.
+The best example is the `nodeId` label.
+We do not want to add the `nodeId` label to the creation of each metric.
+Instead of that we want to add the `nodeId` label to the `Metrics` instance.
+That should happen when a `Metrics` instance is created.
+The creation of a `Metrics` instance is not public api today.
+With the implementation of this proposal a `MetricsFactory` interface should be introduced that allows to create a `Metrics` instance.
+The `MetricsFactory` interface will be defined like this:
+
+```java
+interface MetricsFactory {
+    
+    @NonNull
+    default Metrics create() {
+        return create(Set.of());
+    }
+    
+    @NonNull
+    Metrics create(@NonNull Set<Label> defaultLabels);
+}
+```
+
+The definition of the api should follow our rules regarding services as defined at https://github.com/hashgraph/hedera-services/blob/develop/platform-sdk/docs/base/service-architecture/service-architecture.md
+
+##### Examples
+
+Here are some examples how to define a metric with labels:
+
+```java
+final Label transactionTypeLabel = new Label("transactionType", "fileUpload");
+final Counter.Config config = new Counter.Config("transactionsCategory", "transactionCount").withLabels(transactionTypeLabel);
+
+final Metrics metrics = ...;
+final Counter counter = metrics.getOrCreate(config);
+
+counter.increment();
+```
+
+In prometheus the metric will be defined like this:
+
+```
+transactionsCategory.transactionCount{transactionType="fileUpload"}
+```
+
+#### Snapshot API
+
+The snapshot API will be extended to return the labels of a metric for a snapshot.
+By doing so any monitoring system that uses the snapshot API can use the labels.
+Today we only have 2 system that uses the snapshot API: the prometheus endpoint and the CSV exporter.
+Since no class of the snapshot API is public today (based on the internal usage of `nodeId`) we plan to move parts
+of the snapshot API to the public API.
+
+### Migration
+
+Next to extending the public API we need to migrate the current metrics to the new concept.
+
+#### Migration of the `category`, `name`, and `nodeId`
+
+The `nodeId` information of a metric will be migrated to labels.
```

Comment:
The proposal looks great overall. However, the only part that I cannot find here is how exactly it is proposed to add the new `nodeId` labels given that the `Label`'s value is a `String`. Would the code using this API (or the `MetricsFactory` implementation) just use the `NodeId.toString()`, or there will be a different mechanism to format a `NodeId` as a `String` for the purpose of putting it into a `Label`'s value ? Or perhaps the `Label` could be a generic type like `record Label<T>(String key, T value){}` and just use the `value.toString()` to get the actual string that you use as the value of the label? 

These are just some random options, and personally I don't have any preferences. But I'd expect this document to add some details about the `NodeId` label migration because currently it's unclear.

Reply:
Good point. Today the nodeID is already a label for prometheus. In our prometheus endpoint we therefore migrate a NodeId to a string. The same code will be used in future. The main difference is that this won't happen hidden in the prometheus endpoint code but when creating the `Metrics` instance for a platform.

Reply:
@anthony-swirldslabs do you want me to update the proposal with that information and create a new PR?

Reply:
@hendrikebbers Thanks for clarifying. Given that this update wouldn't be material, I'd be okay with pushing it into this same PR. Another option would be to get this one merged, and then publish a tiny update PR. Or you can leave things as they are, given that you confirmed that you have a clear plan for this particular aspect. I'm good either way. I'll leave it up to you and other reviewers to decide.

Please feel free to resolve this conversation if you have enough approvals and would like to merge the current version of the proposal. Thanks!

Reply:
adding @poulok for feedback to get the process right

Reply:
If you only need to add clarifying comments, i consider that immaterial and can be updated on this PR

Reply:
@anthony-swirldslabs please check my small change here: https://github.com/hashgraph/hedera-services/pull/13975/commits/37df54713e5572047b8647a22cd59e845a4a62c9

Reply:
Looks great. Thank you!